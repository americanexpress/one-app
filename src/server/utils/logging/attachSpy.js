/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export default function attachSpy(obj, methodName, spy) {
  const origMethod = obj[methodName];

  if (typeof origMethod !== 'function') {
    throw new TypeError(`${methodName} is not a function`);
  }

  if (typeof spy !== 'function') {
    throw new TypeError(`spy must be a function (was "${typeof spy}")`);
  }

  // we're monkeypatching, we need to reassign a property of the obj argument
  // eslint-disable-next-line no-param-reassign -- see above comment
  obj[methodName] = function monkeypatchedMethod(...args) {
    let returnValue;
    let originalCalled = false;
    const callOriginal = () => {
      originalCalled = true;
      returnValue = origMethod.apply(obj, args);
      return returnValue;
    };

    spy([...args], callOriginal);

    if (!originalCalled) {
      callOriginal();
    }

    return returnValue;
  };
}
