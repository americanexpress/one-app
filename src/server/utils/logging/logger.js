/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import deepmerge from 'deepmerge';
import { argv } from 'yargs';
import { pino, multistream } from 'pino';
import productionConfig from './config/production';
import otelConfig, {
  createOtelTransport,
} from './config/otel';
import baseConfig from './config/base';

export function createLogger() {
  const useProductionConfig = !!(argv.logFormat === 'machine' || process.env.NODE_ENV !== 'development');

  let transport;

  if (process.env.NODE_ENV === 'development' && process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT) {
    // Temporary solution until https://github.com/Vunovati/pino-opentelemetry-transport/issues/20 is resolved
    // eslint-disable-next-line global-require -- do not load development logger in production
    transport = multistream([{ stream: require('./config/development').default }, createOtelTransport()]);
  } else if (!useProductionConfig && !process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT) {
    // eslint-disable-next-line global-require -- do not load development logger in production
    transport = require('./config/development').default;
  } else if (process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT) {
    transport = createOtelTransport();
  }

  return pino(
    deepmerge(
      baseConfig,
      process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT
        ? otelConfig
        : useProductionConfig && productionConfig
    ),
    transport
  );
}

export default createLogger();
