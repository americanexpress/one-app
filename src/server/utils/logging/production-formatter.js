/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import util from 'util';
import os from 'os';
import dropEntryBasedOnLevel from './level-dropper';
import readJsonFile from '../readJsonFile';

const { buildVersion: version } = readJsonFile('../../../.build-meta.json');
const schemaVersion = '0.3.0';

const application = {
  name: 'One App',
  version,
};

const device = {
  id: `${os.hostname()}:${process.pid}`,
  agent: `${os.type()} ${os.arch()}`,
};

const nodeLevelToSchemaLevel = {
  // 'emergency',
  // 'alert',
  // 'critical',
  error: 'error',
  warn: 'warning',
  log: 'notice',
  info: 'info',
  // 'debug',
};

function generateErrorField(err) {
  return {
    name: err.name,
    // doesn't accept `null`
    message: err.message || '<none>',
    stacktrace: err.stack || '<none>',
    // TODO: parse the stacktrace for the file, line, and column?
    // throwLocation: {
    //   file: 'browser.js',
    //   lineNumber: 1234,
    //   columnNumber: 123,
    //   method: 'functionName',
    // },
  };
}

function getBaseEntry(level) {
  return {
    schemaVersion,
    application,
    device,
    level: nodeLevelToSchemaLevel[level],
    timestamp: new Date().toISOString(),
  };
}

function leadingErrorWithOtherArgs(level, ...args) {
  const [err] = args;
  const entry = getBaseEntry(level);
  entry.error = generateErrorField(err);

  if (err.name === 'ClientReportedError') {
    entry.device = {
      agent: err.userAgent,
    };
    entry.request = {
      address: {
        uri: err.uri,
      },
      metaData: err.metaData,
    };
  }

  if (args.length > 1) {
    const allButFirst = [...args];
    allButFirst.shift();
    entry.message = util.format(...allButFirst);
  }
  return entry;
}

function leadingStringWithError(level, ...args) {
  const entry = getBaseEntry(level);
  entry.error = generateErrorField(args[1]);
  if (args.length === 2) {
    [entry.message] = args;
  } else {
    entry.message = util.format(...args);
  }
  return entry;
}

function entryFromLogType(level, logType) {
  const entry = getBaseEntry(level);
  switch (logType.type) {
    case 'request':
      entry.request = logType.request;
      break;
    default:
      return false;
  }
  return entry;
}

const formatter = (level, ...args) => {
  if (dropEntryBasedOnLevel(level)) {
    return null;
  }

  const [arg0, arg1] = args;

  if (arg0 instanceof Error) {
    return JSON.stringify(leadingErrorWithOtherArgs(level, ...args));
  }
  if (typeof arg0 === 'string' && arg1 instanceof Error) {
    return JSON.stringify(leadingStringWithError(level, ...args));
  }
  if (typeof arg0 === 'object' && arg0 && arg0.type) {
    const entry = entryFromLogType(level, arg0);
    if (entry) {
      return JSON.stringify(entry);
    }
  }

  const entry = getBaseEntry(level);
  entry.message = util.format(...args);
  return JSON.stringify(entry);
};

export default formatter;
