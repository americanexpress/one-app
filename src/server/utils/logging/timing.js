/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// some timing helpers
function startTimer(obj) {
  // attach the data to the request for later use, similar to
  // https://github.com/expressjs/morgan/blob/1533cdcc512f4daf20a93cb0ab3b18d6563a897f/index.js#L491
  /* eslint-disable no-param-reassign, no-underscore-dangle */
  obj._startAt = process.hrtime();
  obj._startTime = new Date().toISOString();
  /* eslint-enable no-param-reassign, no-underscore-dangle */
}

function measureTime(prev) {
  // see `startTimer(...)` for justification
  const previousHrtime = prev._startAt; // eslint-disable-line no-underscore-dangle
  if (!Array.isArray(previousHrtime)) {
    return undefined;
  }
  const diff = process.hrtime(previousHrtime);
  return (diff[0] * 1e3) + (diff[1] * 1e-6); // ms
}

export {
  startTimer,
  measureTime,
};
