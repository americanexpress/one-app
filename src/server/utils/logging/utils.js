/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { STATUS_CODES } from 'http';

import chalk from 'chalk';

const coloredLevels = {
  error: chalk.red('error'),
  warn: chalk.yellow('warn'),
  log: chalk.blue('log'),
  info: chalk.gray('info'),
};

function printStatusCode(obj) {
  const { request: { statusCode } } = obj;

  if (statusCode === null) {
    return chalk.black.bgRed(statusCode);
  }

  if (statusCode < 300) {
    // 100s, 200s
    return chalk.green(statusCode);
  }
  if (statusCode < 400) {
    // 300s
    return chalk.blue(statusCode);
  }
  if (statusCode < 500) {
    // 400s
    return chalk.yellow(statusCode);
  }
  // 500s
  return chalk.red(statusCode);
}

function printStatusMessage(obj) {
  const { request: { statusText, statusCode } } = obj;

  return statusText || STATUS_CODES[statusCode] || chalk.black.bgRed('Timed Out?');
}

function printDurationTime(obj) {
  const { request: { timings: { duration } } } = obj;
  if (duration < 100) {
    return chalk.green(duration);
  }
  if (duration < 1e3) {
    return chalk.yellow(duration);
  }
  if (duration < 6e3) {
    return chalk.red(duration);
  }
  return chalk.black.bgRed(duration);
}

export {
  coloredLevels,
  printStatusCode,
  printStatusMessage,
  printDurationTime,
};
