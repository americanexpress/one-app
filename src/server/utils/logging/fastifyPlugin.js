/*
 * Copyright 2022 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import url from 'url';
import fp from 'fastify-plugin';

import logger from './logger';

/*
TIMERS
`$RequestFullDuration` measures how long it took the whole request to be resolved
`$RequestOverhead` measures how long it took from when the request was made
                   until the route handler was called
`$RouteHandler` measures how long the route handler took to provide an output
`$ResponseBuilder` measures how long it took for fastify to build the response
                   after the payload/output from route handled was provided
Why symbols?
It prevents conflicts with existing keys, native or from other plugins
*/
export const $RequestFullDuration = Symbol('$RequestFullDuration');
export const $RequestOverhead = Symbol('$RequestOverhead');
export const $RouteHandler = Symbol('$RouteHandler');
export const $ResponseBuilder = Symbol('$ResponseBuilder');

const UTILS = {
  configureRequestLog: ({ log }) => log,
};

const getLocale = (req) => {
  // TODO: Verify if `store` is available
  if (req.store) {
    const state = req.store.getState();
    return state.getIn(['intl', 'activeLocale']);
  }

  return undefined;
};

const hrtimeToMs = (value) => (value[0] * 1e3) + (value[1] * 1e-6);

const startTimer = (obj, symbol) => {
  const time = process.hrtime();

  // eslint-disable-next-line no-param-reassign
  obj[symbol] = time;
};

const endTimer = (obj, symbol) => {
  const result = process.hrtime(obj[symbol]);
  const ms = hrtimeToMs(result);

  // eslint-disable-next-line no-param-reassign
  obj[symbol] = ms;

  return ms;
};

const buildMetaData = (request, reply) => {
  const { headers } = request;

  return {
    method: request.method,
    correlationId: headers['correlation-id'],
    // null to explicitly signal no value, undefined if not expected for every request
    host: headers.host || null,
    referrer: headers.referer || headers.referrer || null,
    userAgent: headers['user-agent'] || null,
    location: reply.getHeader('location') || undefined,
    forwarded: headers.forwarded || null,
    forwardedFor: headers['x-forwarded-for'] || null,
    locale: getLocale(request),
  };
};

const logClientRequest = (request, reply) => {
  const getTimer = (symbol) => request[symbol];
  const ttfb = getTimer($RequestFullDuration) - getTimer($ResponseBuilder);
  const log = {
    type: 'request',
    request: {
      direction: 'in',
      protocol: request.protocol,
      address: {
        uri: url.format({
          protocol: request.protocol,
          port: request.port,
          hostname: request.hostname,
          pathname: request.originalUrl,
        }),
      },
      metaData: buildMetaData(request, reply),
      timings: { // https://www.w3.org/TR/navigation-timing/
        // navigationStart: 0,
        // redirectStart: 0,
        // redirectEnd: 0,
        // fetchStart: 0,
        // domainLookupStart: 10,
        // domainLookupEnd: 20,
        // connectStart: 30,
        // secureConnectionStart: 40,
        // connectEnd: 50,
        // requestStart: 60,
        // requestEnd: 70, // optional? not part of the W3C spec
        // responseStart: 80,
        // responseEnd: 90,
        // FIXME: mimic the w3 timing names
        duration: Math.round(getTimer($RequestFullDuration)),
        ttfb: ttfb >= 0 ? Math.round(ttfb) : null,
        requestOverhead: Math.round(getTimer($RequestOverhead)),
        routeHandler: Math.round(getTimer($RouteHandler)),
        responseBuilder: Math.round(getTimer($ResponseBuilder)),
      },
      statusCode: reply.statusCode,
      statusText: reply.raw.statusMessage || undefined,
    },
  };

  const configuredLog = UTILS.configureRequestLog({
    req: request.raw,
    res: reply.raw,
    log,
  });

  logger.info(configuredLog);
};

export const setConfigureRequestLog = (newConfigureRequestLog) => {
  UTILS.configureRequestLog = newConfigureRequestLog;
};

const fastifyPlugin = (fastify, _opts, done) => {
  // decorators should be initialized per Fastify documentation
  // see https://www.fastify.io/docs/latest/Reference/Decorators/#decoratereplyname-value-dependencies
  fastify.decorateRequest($RequestOverhead, null);
  fastify.decorateRequest($RequestFullDuration, null);
  fastify.decorateRequest($RouteHandler, null);
  fastify.decorateRequest($ResponseBuilder, null);

  // NOTE: this is needed for backward compatibility since
  //       we were exposing the 'req' and 'res' from ExpressJS
  //       to the App Config.
  fastify.addHook('onRequest', async (request, reply) => {
    request.raw.originalUrl = request.raw.url;
    request.raw.id = request.id;
    request.raw.hostname = request.hostname;
    request.raw.ip = request.ip;
    request.raw.ips = request.ips;
    request.raw.log = request.log;
    // eslint-disable-next-line no-param-reassign
    reply.raw.log = request.log;

    // backward compatibility for body-parser
    if (request.body) {
      request.raw.body = request.body;
    }
    // backward compatibility for cookie-parser
    if (request.cookies) {
      request.raw.cookies = request.cookies;
    }

    // Make it lazy as it does a bit of work
    Object.defineProperty(request.raw, 'protocol', {
      get() {
        return request.protocol;
      },
    });
  });

  fastify.addHook('onRequest', async (request) => {
    startTimer(request, $RequestOverhead);
    startTimer(request, $RequestFullDuration);
  });

  fastify.addHook('preHandler', async (request) => {
    endTimer(request, $RequestOverhead);
    startTimer(request, $RouteHandler);
  });

  fastify.addHook('onSend', async (request, _reply, payload) => {
    endTimer(request, $RouteHandler);
    startTimer(request, $ResponseBuilder);

    // Note: `onSend` is meant to be used to modify the payload for the response.
    //        we use it to calculate TTFB (time to first byte) since it's called
    //        by fastify right before `writeHead` is called.
    return payload;
  });

  fastify.addHook('onResponse', async (request, reply) => {
    endTimer(request, $ResponseBuilder);
    // same as 'reply.getResponseTime()' but our approach
    // helps us to make the code cleaner
    endTimer(request, $RequestFullDuration);

    logClientRequest(request, reply);
  });

  done();
};

export default fp(fastifyPlugin, {
  fastify: '4.x',
  name: 'logging',
});
