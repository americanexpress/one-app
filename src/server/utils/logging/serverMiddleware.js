/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import url from 'url';

import { monkeypatches } from '@americanexpress/lumberjack';

import { startTimer, measureTime } from './timing';
import logger from './logger';

const passThrough = ({ log }) => log;
const tenantUtils = {
  configureRequestLog: passThrough,
};

function getLocale(req) {
  if (req.store) {
    const state = req.store.getState();
    return state.getIn(['intl', 'activeLocale']);
  }
  return undefined;
}

function buildMetaData(req, res) {
  const { headers } = req;
  return {
    method: req.method,
    correlationId: headers['correlation-id'],
    // null to explicitly signal no value, undefined if not expected for every request
    host: headers.host || null,
    referrer: headers.referer || headers.referrer || null,
    userAgent: headers['user-agent'] || null,
    location: res.getHeader('location') || undefined,
    forwarded: headers.forwarded || null,
    forwardedFor: headers['x-forwarded-for'] || null,
    locale: getLocale(req),
  };
}

function logClientRequest(req, res) {
  const fullDuration = measureTime(req);
  const responseDuration = measureTime(res);
  const ttfb = fullDuration - responseDuration;
  const log = {
    type: 'request',
    request: {
      direction: 'in',
      protocol: req.protocol,
      address: {
        uri: url.format({
          protocol: req.protocol,
          port: req.port,
          hostname: req.hostname,
          pathname: req.originalUrl,
        }),
      },
      metaData: buildMetaData(req, res),
      timings: { // https://www.w3.org/TR/navigation-timing/
        // navigationStart: 0,
        // redirectStart: 0,
        // redirectEnd: 0,
        // fetchStart: 0,
        // domainLookupStart: 10,
        // domainLookupEnd: 20,
        // connectStart: 30,
        // secureConnectionStart: 40,
        // connectEnd: 50,
        // requestStart: 60,
        // requestEnd: 70, // optional? not part of the W3C spec
        // responseStart: 80,
        // responseEnd: 90,
        // FIXME: mimic the w3 timing names
        duration: Math.round(fullDuration),
        ttfb: ttfb >= 0 ? Math.round(ttfb) : null,
      },
      statusCode: res.statusCode,
      statusText: res.statusMessage || undefined,
    },
  };

  const configuredLog = tenantUtils.configureRequestLog({ req, res, log });
  logger.info(configuredLog);
}

export const setConfigureRequestLog = (newConfigureRequestLog = passThrough) => {
  tenantUtils.configureRequestLog = newConfigureRequestLog;
};

export default function serverMiddleware(req, res, next) {
  startTimer(req);
  monkeypatches.attachSpy(res, 'writeHead', () => startTimer(res));
  res.on('close', () => logClientRequest(req, res));
  setImmediate(next);
}
