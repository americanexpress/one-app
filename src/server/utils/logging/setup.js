/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { monkeypatches } from '@americanexpress/lumberjack';

import lumberjackLogger from './logger';
import { replaceGlobalConsoleWithOtelLogger } from './otel/logger';
import { startTimer, measureTime } from './timing';

const COLON_AT_THE_END_REGEXP = /:$/;
function formatProtocol(parsedUrl) {
  const { protocol } = parsedUrl;
  return protocol.replace(COLON_AT_THE_END_REGEXP, '');
}

let logger;

if (process.env.OTEL_EXPORTER_OTLP_LOGS_ENDPOINT) {
  logger = replaceGlobalConsoleWithOtelLogger();
} else {
  logger = lumberjackLogger;
  monkeypatches.replaceGlobalConsole(logger);
}

function outgoingRequestSpy(externalRequest) {
  startTimer(externalRequest);
}

function outgoingRequestEndSpy(externalRequest, parsedUrl) {
  const { res } = externalRequest;
  const duration = Math.round(measureTime(externalRequest));
  logger.info({
    type: 'request',
    request: {
      direction: 'out',
      protocol: formatProtocol(parsedUrl),
      address: {
        uri: parsedUrl.href,
      },
      metaData: {
        method: externalRequest.method,
        // null to explicitly signal no value, undefined if not expected for every request
        correlationId: externalRequest.getHeader('correlation-id') || undefined,
      },
      timings: { // https://www.w3.org/TR/navigation-timing/
        // navigationStart: 0,
        // redirectStart: 0,
        // redirectEnd: 0,
        // fetchStart: 0,
        // domainLookupStart: 10,
        // domainLookupEnd: 20,
        // connectStart: 30,
        // secureConnectionStart: 40,
        // connectEnd: 50,
        // requestStart: 60,
        // requestEnd: 70, // optional? not part of the W3C spec
        // responseStart: 80,
        // responseEnd: 90,
        // FIXME: mimic the w3 timing names
        duration,
      },
      statusCode: (res && res.statusCode) || null,
      statusText: (res && res.statusMessage) || null,
    },
  });
}

// In Node.js v8 and earlier https.request internally called http.request, but this is changed in
// later versions
// https://github.com/nodejs/node/blob/v6.x/lib/https.js#L206
// https://github.com/nodejs/node/blob/v8.x/lib/https.js#L239
// https://github.com/nodejs/node/blob/v10.x/lib/https.js#L271
if (Number.parseInt(/^v(\d+)/.exec(process.version)[1], 10) > 8) {
  monkeypatches.attachHttpsRequestSpy(outgoingRequestSpy, outgoingRequestEndSpy);
}
monkeypatches.attachHttpRequestSpy(outgoingRequestSpy, outgoingRequestEndSpy);
