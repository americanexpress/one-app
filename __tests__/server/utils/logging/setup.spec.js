/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import url from 'url';

jest.mock('yargs', () => ({
  argv: {
    logLevel: 'debug',
  },
}));

describe('setup', () => {
  let attachRequestSpies;
  let logger;
  let startTimer;

  const logMethods = ['error', 'warn', 'log', 'info', 'debug'];
  const originalConsole = logMethods.reduce((acc, curr) => ({
    ...acc,
    [curr]: console[curr],
  }), {});

  afterEach(() => {
    logMethods.forEach((method) => { console[method] = originalConsole[method]; });
  });

  function load() {
    jest.resetModules();

    jest.mock('../../../../src/server/utils/logging/attachRequestSpies');
    jest.mock('../../../../src/server/utils/logging/timing', () => ({
      startTimer: jest.fn(),
      measureTime: jest.fn(() => 12),
    }));
    attachRequestSpies = require('../../../../src/server/utils/logging/attachRequestSpies').default;
    ({ startTimer } = require('../../../../src/server/utils/logging/timing'));
    logger = require('../../../../src/server/utils/logging/logger').default;
    require('../../../../src/server/utils/logging/setup');
    jest.spyOn(logger, 'info').mockImplementation(() => 0);
  }

  it('replaces the global console with logger', () => {
    load();
    logMethods.forEach((method) => {
      expect(console[method].name).toBe('bound hookWrappedLog');
      expect(console[method]).not.toBe(originalConsole[method]);
    });
  });

  it('spies on requests', () => {
    load();
    expect(attachRequestSpies).toHaveBeenCalledTimes(1);
  });

  describe('logging outgoing requests', () => {
    function createExternalRequestAndParsedUrl() {
      const externalRequestHeaders = {
        'correlation-id': '123',
        host: 'example.com',
        referer: 'https://example.com/other-place',
        'user-agent': 'Browser/8.0 (compatible; WXYZ 100.0; Doors LQ 81.4; Boat/1.0)',
      };
      const externalRequest = {
        method: 'GET',
        getHeader: jest.fn((name) => externalRequestHeaders[name]),
        res: {
          statusCode: 200,
          statusMessage: 'OK',
        },
      };
      const parsedUrl = url.parse('https://example.com/place');
      return { externalRequest, parsedUrl, externalRequestHeaders };
    }

    it('starts a timer when the request starts', () => {
      const { externalRequest } = createExternalRequestAndParsedUrl();
      load();
      expect(attachRequestSpies).toHaveBeenCalledTimes(1);
      const outgoingRequestSpy = attachRequestSpies.mock.calls[0][0];
      outgoingRequestSpy(externalRequest);
      expect(startTimer).toHaveBeenCalledTimes(1);
      expect(startTimer).toHaveBeenCalledWith(externalRequest);
    });

    it('is level info', () => {
      const { externalRequest, parsedUrl } = createExternalRequestAndParsedUrl();
      load();
      expect(attachRequestSpies).toHaveBeenCalledTimes(1);
      const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
      outgoingRequestEndSpy(externalRequest, parsedUrl);
      expect(logger.info).toHaveBeenCalledTimes(1);
      expect(logger.info.mock.calls[0]).toMatchSnapshot();
    });

    it('uses the correlation id header if present', () => {
      const {
        externalRequest,
        externalRequestHeaders,
        parsedUrl,
      } = createExternalRequestAndParsedUrl();
      externalRequestHeaders['correlation-id'] = '1234';
      load();
      expect(attachRequestSpies).toHaveBeenCalledTimes(1);
      const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
      outgoingRequestEndSpy(externalRequest, parsedUrl);
      expect(logger.info).toHaveBeenCalledTimes(1);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('correlationId', '1234');
    });

    it('uses undefined for the correlation id header if not present', () => {
      // TODO: add holocron API to edit the request so we can add tracing/correlationId headers
      const {
        externalRequest,
        externalRequestHeaders,
        parsedUrl,
      } = createExternalRequestAndParsedUrl();
      delete externalRequestHeaders['correlation-id'];
      load();
      expect(attachRequestSpies).toHaveBeenCalledTimes(1);
      const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
      outgoingRequestEndSpy(externalRequest, parsedUrl);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request.metaData).toHaveProperty('correlationId', undefined);
    });

    it('uses the status code if present', () => {
      const { externalRequest, parsedUrl } = createExternalRequestAndParsedUrl();
      externalRequest.res.statusCode = 200;
      load();
      expect(attachRequestSpies).toHaveBeenCalledTimes(1);
      const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
      outgoingRequestEndSpy(externalRequest, parsedUrl);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request).toHaveProperty('statusCode', 200);
    });

    it('uses null for the status code if not present', () => {
      const { externalRequest, parsedUrl } = createExternalRequestAndParsedUrl();
      delete externalRequest.res.statusCode;
      load();
      expect(attachRequestSpies).toHaveBeenCalledTimes(1);
      const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
      outgoingRequestEndSpy(externalRequest, parsedUrl);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request).toHaveProperty('statusCode', null);
    });

    it('uses the status text if present', () => {
      const { externalRequest, parsedUrl } = createExternalRequestAndParsedUrl();
      externalRequest.res.statusMessage = 'OK';
      load();
      expect(attachRequestSpies).toHaveBeenCalledTimes(1);
      const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
      outgoingRequestEndSpy(externalRequest, parsedUrl);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request).toHaveProperty('statusText', 'OK');
    });

    it('uses null for the status text if not present', () => {
      const { externalRequest, parsedUrl } = createExternalRequestAndParsedUrl();
      delete externalRequest.res.statusMessage;
      load();
      expect(attachRequestSpies).toHaveBeenCalledTimes(1);
      const outgoingRequestEndSpy = attachRequestSpies.mock.calls[0][1];
      outgoingRequestEndSpy(externalRequest, parsedUrl);
      const entry = logger.info.mock.calls[0][0];
      expect(entry.request).toHaveProperty('statusText', null);
    });
  });
});
