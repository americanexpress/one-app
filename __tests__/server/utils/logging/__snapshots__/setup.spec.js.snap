// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setup logging outgoing requests is level info 1`] = `
[
  {
    "request": {
      "address": {
        "uri": "https://example.com/place",
      },
      "direction": "out",
      "metaData": {
        "correlationId": "123",
        "method": "GET",
      },
      "protocol": "https",
      "statusCode": 200,
      "statusText": "OK",
      "timings": {
        "duration": 12,
      },
    },
  },
]
`;
